pkg/jobs/job_scheduler.go

```diff
@@ -254,23 +288,40 @@ func (s *jobScheduler) executeSchedules(
 			continue
 		}
 
-		sp, err := txn.CreateSavepoint(ctx)
-		if err != nil {
-			return err
-		}
-
-		if err := s.processSchedule(ctx, schedule, numRunning, stats, txn); err != nil {
-			log.Errorf(ctx, "error processing schedule %d: %+v", schedule.ScheduleID(), err)
-
-			if err := txn.RollbackToSavepoint(ctx, sp); err != nil {
-				return errors.Wrapf(err, "failed to rollback savepoint for schedule %d", schedule.ScheduleID())
+		if processErr := withSavePoint(ctx, txn, func() error {
+			return s.processSchedule(ctx, schedule, numRunning, stats, txn)
+		}); processErr != nil {
+			if errors.HasType(err, (*savePointError)(nil)) {
+				return errors.Wrapf(err, "savepoint error for schedule %d", schedule.ScheduleID())
 			}
-		}
 
-		if err := txn.ReleaseSavepoint(ctx, sp); err != nil {
-			return err
+			// Failed to process schedule.
+			s.metrics.NumBadSchedules.Inc(1)
+			log.Errorf(ctx,
+				"error processing schedule %d: %+v", schedule.ScheduleID(), processErr)
+
+			// Try updating schedule record to indicate schedule execution error.
+			// We can't be sure that processSchedule didn't mutate schedule.
```

Comment:
Eh, now this is why I don't like this ORM pattern -- here we need to load our object to act on it when all what we want is a single-roundtrip `executor.Exec(ctx, "UPDATE %s SET status =...)`. I know everything else about schedule uses its little ORM instead of using SQL so it makes sense to keep doing that here though so not suggesting otherwise, just have an axe to grind.